src/server/db_slice.cc

```diff
@@ -816,42 +827,45 @@ OpResult<int64_t> DbSlice::UpdateExpire(const Context& cntx, PrimeIterator prime
   }
 }
 
-std::pair<PrimeIterator, bool> DbSlice::AddOrUpdateInternal(const Context& cntx,
-                                                            std::string_view key, PrimeValue obj,
-                                                            uint64_t expire_at_ms,
-                                                            bool force_update) noexcept(false) {
+DbSlice::AddOrFindResult DbSlice::AddOrUpdateInternal(const Context& cntx, std::string_view key,
+                                                      PrimeValue obj, uint64_t expire_at_ms,
+                                                      bool force_update) noexcept(false) {
   DCHECK(!obj.IsRef());
 
-  pair<PrimeIterator, bool> res = AddOrFind(cntx, key);
-  if (!res.second && !force_update)  // have not inserted.
+  auto res = AddOrFind(cntx, key);
+  if (!res.is_new && !force_update)  // have not inserted.
     return res;
 
   auto& db = *db_arr_[cntx.db_index];
-  auto& it = res.first;
+  auto& it = res.it;
 
   it->second = std::move(obj);
-  PostUpdate(cntx.db_index, it, key, false);
 
   if (expire_at_ms) {
     it->second.SetExpire(true);
     uint64_t delta = expire_at_ms - expire_base_[0];
-    auto [eit, inserted] = db.expire.Insert(it->first.AsRef(), ExpirePeriod(delta));
-    CHECK(inserted || force_update);
-    if (!inserted) {
-      eit->second = ExpirePeriod(delta);
+    if (IsValid(res.exp_it) && force_update) {
+      res.exp_it->second = ExpirePeriod(delta);
+    } else {
+      auto [eit, inserted] = db.expire.Insert(it->first.AsRef(), ExpirePeriod(delta));
+      CHECK(inserted || force_update);
```

Comment:
I think this should be only CHECK(inserted), as if it was already in the expire db it should be in the above if

Reply:
I'll use the new `InsertNew()` instead then

Reply:
Wait, what about the case of `AddOrSkip()`?
If `res.exp_it` is valid (exists), but `force_update` is false? Then it's valid for `Insert()` to return `inserted == false`, no?

Reply:
we should check the flows that calls AddOrSkip. I am not sure if force_update is false should it update the expire time

Reply:
Unless I'm missing something, the only call we have is in `RdbRestoreValue::Add` - what is the expected behavior there?

Reply:
I actually think that RdbRestoreValue::Add should call AddNew as if restore command was called with parameter to overide existing entry re first delete the entry

Reply:
Agreed. I just checked, and indeed the `RESTORE` flow first checks for existence (and removes if needed).
So I removed `AddOrSkip()` entirely.